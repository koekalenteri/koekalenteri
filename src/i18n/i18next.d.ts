import type { fi, fiBreed, fiBreedAbbr, fiCountry } from './locales'

import 'i18next'

declare module 'i18next' {
  interface CustomTypeOptions {
    resources: {
      translation: typeof fi
      breed: typeof fiBreed
      breedAbbr: typeof fiBreedAbbr
      country: typeof fiCountry
    }
  }
}
