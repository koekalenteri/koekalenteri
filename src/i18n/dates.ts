import type { Language } from '../types'

import { formatDistanceToNowStrict, isSameDay, isSameMonth, isSameYear, isValid, parseISO } from 'date-fns'
import en from 'date-fns/locale/en-GB/index'
import fi from 'date-fns/locale/fi/index'
import { formatInTimeZone } from 'date-fns-tz'

export interface DateFormatOptions {
  tz?: string
  locale?: Locale
}

export const locales: Record<Language, Locale> = { en, fi }

export function formatDate(
  date: Date | string,
  format: string,
  { tz = 'Europe/Helsinki', locale = fi }: DateFormatOptions = {}
): string {
  if (typeof date === 'string') {
    date = parseISO(date)
  }

  if (!isValid(date)) {
    return ''
  }

  return formatInTimeZone(date, tz, format, { locale })
}

export const getDateFormatter =
  (fmt: string) =>
  (date: Date | string, lng: string | undefined): string => {
    const locale = locales[lng as Language]

    return formatDate(date, fmt, { locale })
  }

export function formatDateSpan(
  start: Date | string,
  lng: string | undefined,
  { end, noYear }: { end: Date | string; noYear?: boolean },
  tz = 'Europe/Helsinki'
): string {
  const y = noYear ? '' : 'yyyy'
  if (typeof start === 'string') {
    start = parseISO(start)
  }
  if (typeof end === 'string') {
    end = parseISO(end)
  }
  if (!isValid(start)) {
    return ''
  }
  if (!isValid(end)) {
    end = start
  }
  if (isSameDay(start, end)) {
    return formatInTimeZone(start, tz, `d.M.${y}`)
  }
  if (isSameMonth(start, end)) {
    return formatInTimeZone(start, tz, 'd.') + '–' + formatInTimeZone(end, tz, `d.M.${y}`)
  }
  if (isSameYear(start, end)) {
    return formatInTimeZone(start, tz, 'd.M.') + '–' + formatInTimeZone(end, tz, `d.M.${y}`)
  }
  return formatInTimeZone(start, tz, `d.M.${y}`) + '–' + formatInTimeZone(end, tz, `d.M.${y}`)
}

export function formatDistance(date?: Date, lng?: string): string {
  const locale = locales[lng as Language]
  return formatDistanceToNowStrict(date ?? new Date(), { locale })
}

export const currentFinnishTime = (): string => formatDate(new Date(), "yyyy-MM-dd'T'HH:mm:ssxxx")
