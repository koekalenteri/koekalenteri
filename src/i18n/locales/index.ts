export { default as fi } from './fi/translation.json'
export { default as en } from './en/translation.json'

export { default as fiBreed } from './fi/breed.json'
export { default as enBreed } from './en/breed.json'

export { default as fiBreedAbbr } from './fi/breed-abbr.json'
export { default as enBreedAbbr } from './en/breed-abbr.json'
