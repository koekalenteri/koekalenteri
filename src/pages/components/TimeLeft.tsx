import { useTranslation } from 'react-i18next'
import Typography from '@mui/material/Typography'

interface Props {
  date?: Date
}

export const TimeLeft = ({ date }: Props) => {
  const { t } = useTranslation()

  if (!date) return null

  return (
    <Typography display="inline" ml={1}>
      {t('dateFormat.distanceLeft', { date })}
    </Typography>
  )
}
