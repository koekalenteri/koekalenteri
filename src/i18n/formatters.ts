import type { i18n } from 'i18next'

import { formatDateSpan, formatDistance, getDateFormatter } from './dates'

export const registerFormatters = (i18n: i18n) => {
  i18n.services.formatter?.add('dtshort', getDateFormatter('eeeeee d.M. HH:mm'))
  i18n.services.formatter?.add('wdshort', getDateFormatter('eeeeee d.M.'))
  i18n.services.formatter?.add('weekday', getDateFormatter('eeeeee'))
  i18n.services.formatter?.add('short', getDateFormatter('dd.MM.'))
  i18n.services.formatter?.add('datespan', formatDateSpan)
  i18n.services.formatter?.add('distance', formatDistance)
  i18n.services.formatter?.add('lowercase', (value) => value.toLowerCase())
  i18n.services.formatter?.add('date', getDateFormatter('d.M.yyyy'))
  i18n.services.formatter?.add('isodate', getDateFormatter('dd.MM.yyyy'))
  i18n.services.formatter?.add('long', getDateFormatter('dd.MM.yyyy HH:mm'))
}
