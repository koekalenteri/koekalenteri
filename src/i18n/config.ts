import type { InitOptions } from 'i18next'

import { en, enBreed, enBreedAbbr, enCountry, fi, fiBreed, fiBreedAbbr, fiCountry } from './locales'

export const i18nInit: InitOptions = {
  lng: 'fi',
  ns: ['translation', 'breed', 'breedAbbr', 'country'],
  resources: {
    fi: { translation: fi, breed: fiBreed, breedAbbr: fiBreedAbbr, country: fiCountry },
    en: { translation: en, breed: enBreed, breedAbbr: enBreedAbbr, country: enCountry },
  },
  fallbackLng: 'fi',
  supportedLngs: ['fi', 'en'],
  debug: process.env.NODE_ENV === 'development',
  interpolation: {
    escapeValue: false,
  },
}
