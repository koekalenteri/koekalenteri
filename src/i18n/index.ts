import type { Localization } from '@mui/material/locale'
import type { Language } from '../types'
import type { fi } from './locales'

import { initReactI18next } from 'react-i18next'
import { enUS, fiFI } from '@mui/material/locale'
import { enUS as genUS, fiFI as gfiFI } from '@mui/x-data-grid/locales'
import { enUS as denUS, fiFI as dfiFI } from '@mui/x-date-pickers/locales'
import i18n from 'i18next'

import { i18nInit } from './config'
import { locales } from './dates'
import { registerFormatters } from './formatters'

export { locales }
export type { Language }
export type ValidationErrorKey = typeof fi.validation

export const muiLocales: Record<Language, Localization> = {
  fi: { ...fiFI, components: { ...fiFI.components, ...gfiFI.components, ...dfiFI.components } },
  en: { ...enUS, components: { ...enUS.components, ...genUS.components, ...denUS.components } },
}

i18n
  .use(initReactI18next)
  .init(i18nInit)
  .catch((reason) => console.error(reason))

//  additional formats
registerFormatters(i18n)
