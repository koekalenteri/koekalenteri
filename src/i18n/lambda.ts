import type { Language } from '../types'

import i18n from 'i18next'

import { formatDateSpan, formatDistance, getDateFormatter, locales } from './dates'
import { en, enBreed, enBreedAbbr, fi, fiBreed, fiBreedAbbr } from './locales'

export { locales }
export { i18n }
export type { Language }
export type ValidationErrorKey = typeof fi.validation

i18n
  .init({
    lng: 'fi',
    ns: ['translation', 'breed', 'breedAbbr'],
    resources: {
      fi: { translation: fi, breed: fiBreed, breedAbbr: fiBreedAbbr },
      en: { translation: en, breed: enBreed, breedAbbr: enBreedAbbr },
    },
    fallbackLng: 'fi',
    supportedLngs: ['fi', 'en'],
    debug: process.env.NODE_ENV === 'development',
    interpolation: {
      escapeValue: false,
    },
  })
  .catch((reason) => console.error(reason))

//  additional formats
i18n.services.formatter?.add('dtshort', getDateFormatter('eeeeee d.M. HH:mm'))
i18n.services.formatter?.add('wdshort', getDateFormatter('eeeeee d.M.'))
i18n.services.formatter?.add('weekday', getDateFormatter('eeeeee'))
i18n.services.formatter?.add('short', getDateFormatter('dd.MM.'))
i18n.services.formatter?.add('datespan', formatDateSpan)
i18n.services.formatter?.add('distance', formatDistance)
i18n.services.formatter?.add('lowercase', (value) => value.toLowerCase())
i18n.services.formatter?.add('date', getDateFormatter('d.M.yyyy'))
i18n.services.formatter?.add('isodate', getDateFormatter('dd.MM.yyyy'))
i18n.services.formatter?.add('long', getDateFormatter('dd.MM.yyyy HH:mm'))
