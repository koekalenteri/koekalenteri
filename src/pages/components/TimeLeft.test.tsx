import type { ReactNode } from 'react'

import { ThemeProvider } from '@mui/material'
import { LocalizationProvider } from '@mui/x-date-pickers'
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFnsV3'
import { render, screen } from '@testing-library/react'

import theme from '../../assets/Theme'
import { locales } from '../../i18n'

import { TimeLeft } from './TimeLeft'

function Wrapper({ children }: { readonly children: ReactNode }) {
  return (
    <ThemeProvider theme={theme}>
      <LocalizationProvider dateAdapter={AdapterDateFns} adapterLocale={locales.fi}>
        {children}
      </LocalizationProvider>
    </ThemeProvider>
  )
}

describe('TimeLeft', () => {
  it('renders with date', () => {
    const now = new Date()
    const { container } = render(<TimeLeft date={now} />, { wrapper: Wrapper })

    expect(container.firstChild).toMatchSnapshot()
    expect(screen.getByText('dateFormat.distanceLeft date')).toBeInTheDocument()
  })

  it('renders without date', () => {
    const { container } = render(<TimeLeft />, { wrapper: Wrapper })

    expect(container.firstChild).toMatchSnapshot()
    expect(container.firstChild).toBeNull()
  })
})
