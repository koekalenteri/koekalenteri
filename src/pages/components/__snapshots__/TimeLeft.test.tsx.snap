// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeLeft renders with date 1`] = `
<p
  class="MuiTypography-root MuiTypography-body1 css-150xto-MuiTypography-root"
>
  dateFormat.distanceLeft date
</p>
`;

exports[`TimeLeft renders without date 1`] = `null`;
